source/server/connection_handler_impl.cc

```diff
@@ -293,54 +314,89 @@ void ConnectionHandlerImpl::ActiveTcpListener::newConnection(
   if (empty_filter_chain) {
-    ENVOY_CONN_LOG(debug, "closing connection: no filters", *new_connection);
-    new_connection->close(Network::ConnectionCloseType::NoFlush);
-    return;
+    ENVOY_CONN_LOG(debug, "closing connection: no filters", *active_connection->connection_);
+    active_connection->connection_->close(Network::ConnectionCloseType::NoFlush);
   }
 
-  onNewConnection(std::move(new_connection));
-}
-
-void ConnectionHandlerImpl::ActiveTcpListener::onNewConnection(
-    Network::ConnectionPtr&& new_connection) {
-  ENVOY_CONN_LOG(debug, "new connection", *new_connection);
-
   // If the connection is already closed, we can just let this connection immediately die.
-  if (new_connection->state() != Network::Connection::State::Closed) {
-    ActiveConnectionPtr active_connection(
-        new ActiveConnection(*this, std::move(new_connection), parent_.dispatcher_.timeSource()));
+  if (active_connection->connection_->state() != Network::Connection::State::Closed) {
+    ENVOY_CONN_LOG(debug, "new connection", *active_connection->connection_);
+    active_connection->connection_->addConnectionCallbacks(*active_connection);
     active_connection->moveIntoList(std::move(active_connection), connections_);
-    parent_.num_connections_++;
   }
 }
 
-ConnectionHandlerImpl::ActiveConnection::ActiveConnection(ActiveTcpListener& listener,
-                                                          Network::ConnectionPtr&& new_connection,
-                                                          TimeSource& time_source)
+namespace {
+// Structure used to allow a unique_ptr to be captured in a posted lambda. See below.
+struct RebalancedSocket {
+  Network::ConnectionSocketPtr socket;
+};
+using RebalancedSocketSharedPtr = std::shared_ptr<RebalancedSocket>;
+} // namespace
+
+void ConnectionHandlerImpl::ActiveTcpListener::post(Network::ConnectionSocketPtr&& socket) {
+  // It is not possible to capture a unique_ptr because the post() API copies the lambda, so we must
+  // bundle the socket inside a shared_ptr that can be captured.
+  // TODO(mattklein123): It may be possible to change the post() API such that the lambda is only
+  // moved, but this is non-trivial and needs investigation.
+  RebalancedSocketSharedPtr socket_to_rebalance = std::make_shared<RebalancedSocket>();
+  socket_to_rebalance->socket = std::move(socket);
+
+  parent_.dispatcher_.post([socket_to_rebalance, tag = config_.listenerTag(), &parent = parent_]() {
+    // TODO(mattklein123): We should probably use a hash table here to lookup the tag instead of
+    // iterating through the listener list.
+    for (const auto& listener : parent.listeners_) {
+      if (listener.second->listener() != nullptr && listener.second->listenerTag() == tag) {
+        // TODO(mattklein123): Remove dynamic_cast here in follow up cleanup.
+        ActiveTcpListener& tcp_listener = *dynamic_cast<ActiveTcpListener*>(listener.second.get());
```

Comment:
suggest:
```
ActiveTcpListener* tcp_listener = dynamic_cast<ActiveTcpListener*>(listener.second.get());
ASSERT(tcp_listener);
tcp_listener->onAccept ....
```
so that you get a more useful error if the dynamic cast fails in fastbuild/debug.

Reply:
I will skip this one as I think it will still crash in an obvious way and I'm going to do a follow up refactor to remove the dynamic casts anyway.

Reply:
ok the other one is just a typo so feel free to do that in your follow-up too, so we don't have to re-wait for CI.

Reply:
too late :)